### ISSUE TITLE
__builtin_dump_struct buffer overrun with types it doesn't handle

### ISSUE THREAD

**Issue by ndkrempel** [llvm/llvm-project#44489]
|  |  |
| --- | --- |
| Bugzilla Link | [45144](https://llvm.org/bz45144) |
| Version | 9.0 |
| OS | All |
| CC | @DougGregor,@zygoloid |

## Extended Description 
The following program:

```C++
#include <cstdio>
struct Foo {
    short a[1];
};
int main() {
    Foo f = {2};
    __builtin_dump_struct(&f, &std::printf);
}
```

prints out garbage data from clang's memory:

```
struct Foo {
short [1] a : 0x315b2074726f0002
}
```

The `0002` on the end is the valid data, the rest is presumably whatever happened to be in the compiler's memory afterwards.

A similar thing happens with any field type that `__builtin_dump_struct` doesn't understand - it prints out 8 bytes (or maybe 4 bytes on a 32-bit system) of data, regardless of the actual size of the field.

Reproduced on clang 9.0.0 and trunk. Didn't seem to occur on 7.0.0 and 8.0.0 but that could have just been getting lucky that the extra bytes displayed were all 0.

**Comment by ndkrempel** [comment]
This seems to have been fixed starting in clang version 15.0.0, where arrays are output just as their address with a `*` prefix.

**Comment by tbaederr** [comment]
Output is now
```
Foo {
  short[1] a = *0x7ffef463093e
}

```
